src/crimson/admin/osd_admin.cc

```diff
@@ -0,0 +1,319 @@
+	return seastar::make_ready_future<tell_result_t>(
+	  tell_result_t{-EINVAL, "syntax error: 'config set <var> <value>'"});
+      }
+      // val may be multiple words
+      string valstr = str_join(new_val, " ");
+      return local_conf().set_val(var, valstr).then([f] {
+	f->open_object_section("config_set");
+        f->dump_string("success", "");
+	f->close_section();
+	return seastar::make_ready_future<tell_result_t>();
+      }).handle_exception_type([](std::invalid_argument& e) {
+	return seastar::make_ready_future<tell_result_t>(
+          tell_result_t{-EINVAL, e.what()});
+      });
+    }
+  };
+
+  /**
+   * A CephContext admin hook: calling assert (if allowed by
+   * 'debug_asok_assert_abort')
+   */
+  class AssertAlwaysHook : public OsdAdminHookBase {
+   public:
+    explicit AssertAlwaysHook(OsdAdminImp& master)
+        : OsdAdminHookBase(master) {}
+    seastar::future<tell_result_t> tell(const cmdmap_t&,
+					ceph::Formatter*) const final
+    {
+      if (local_conf().get_val<bool>("debug_asok_assert_abort")) {
+	ceph_assert_always(0);
+	return seastar::make_ready_future<tell_result_t>();
+      } else {
+	return seastar::make_ready_future<tell_result_t>(
+	  tell_result_t{-EPERM, "configuration set to disallow asok assert"});
+      }
+    }
+  };
+
+  /**
+   * provide the hooks with access to OSD internals
+   */
+  const OSDSuperblock& osd_superblock() const
+  {
+    return m_osd->superblock;
+  }
+
+  OSDState get_osd_state() const
+  {
+    return m_osd->state;
+  }
+
+  string_view osd_state_name() const
+  {
+    return m_osd->state.to_string();
+  }
+
+  OsdStatusHook osd_status_hook{*this};
+  SendBeaconHook send_beacon_hook{*this};
+  ConfigShowHook config_show_hook{*this};
+  ConfigGetHook config_get_hook{*this};
+  ConfigSetHook config_set_hook{*this};
+  AssertAlwaysHook assert_hook{*this};
+
+ public:
+  OsdAdminImp(OSD* osd)
+      : m_osd{ osd }
+  {}
+
+  ~OsdAdminImp() = default;
+
+  seastar::future<> register_admin_commands()
+  {
+    static const std::vector<AsokServiceDef> hooks_tbl{
+      // clang-format off
+        AsokServiceDef{"status",      "status",        &osd_status_hook,
+                      "OSD status"}
+      , AsokServiceDef{"send_beacon", "send_beacon",   &send_beacon_hook,
+                      "send OSD beacon to mon immediately"}
+      , AsokServiceDef{"config show", "config show", &config_show_hook,
+                      "dump current config settings" }
+      , AsokServiceDef{"config get", "config get name=var,type=CephString",
+		       &config_get_hook,
+		       "config get <field>: get the config value" }
+      , AsokServiceDef{"config set",
+		       "config set name=var,type=CephString name=val,type=CephString,n=N",
+		       &config_set_hook,
+		       "config set <field> <val> [<val> ...]: set a config variable" }
+      , AsokServiceDef{ "assert", "assert", &assert_hook, "asserts" }
+      // clang-format on
+    };
+
+    return m_osd->asok
```

Comment:
i'd avoid accessing osd's private member variable in this way. would better to do  register the hooks in `AdminSocket::start()`.

but we can certainly leave this to a follow-up PR. if you are fine with it, i can help on this.

Reply:
We better discuss this issue separately. I can see a few arguments against this specific change.